### DISCUSSION TITLE
Problems with Merge Confidence PR styling

### DISCUSSION THREAD

**Discussion by HonkingGoose** [whitesource/merge-confidence#10] [General]
## Problem:

The badges for _Age_ and _Passing_ shrink down to unreadable sizes when the viewport is made smaller.
This affects the New and Low confidences badges, and might affect other badges in the range as well.

### Full viewport:
![merge-confidence-full-size-PR](https://user-images.githubusercontent.com/34918129/100781648-94c2a180-340b-11eb-9f15-a16d4dac750e.png)

### Small viewport:
![merge-confidence-small-size-PR](https://user-images.githubusercontent.com/34918129/100781656-97bd9200-340b-11eb-8e7c-57f258ef0fe3.png)

## How to reproduce:

1. Go to a PR which has merge confidence enabled.
1. Look at PR with a full width viewport, note all badges readable.
1. Shrink down viewport width, note _Age_ and _Passing_ badges become way too small.

## PR where these pictures are coming from:

https://github.com/RoostingGeese/git-gosling/pull/94

**Comment by rarkins** [comment]
Is this possible to *fix* though? The browser appears to be adapting based on a combination of column header with / badge width.

**Comment by HonkingGoose** [comment]
> The browser appears to be adapting based on a combination of column header with / badge width.

I've tested this, badge size is based on column header size.

Reproduction:

1. Open PR in browser.
2. Open dev-tools.
3. Adjust table headers to only use one character per table header.
4. Play with viewport size.
5. Note that all badges shrink down to really tiny sizes.

**Comment by rarkins** [comment]
Given a small enough viewpoint, won't all badges shrink? Or there is there some minimum based on the column header width?

**Comment by HonkingGoose** [comment]
There is a minimum width based on the column header width.

## Badges more or less keep normal size with larger column text:

![badges-keep-size-better](https://user-images.githubusercontent.com/34918129/100885950-83cd6b00-34b3-11eb-9588-f2b378c878f0.png)

---

I also noticed something else when I was changing the column header text.
The longer text exposed that the badges are not centered in their columns.

You can use ` :---:` instead of `---` in the GitHub Flavoured Markdown tables to center the badges horizontally:

```markdown
| Package | Change | Age | Adoption | Passing | Confidence |
| :---: | :---: | :---: | :---: | :---: | :---: |
| [@mdx-js/react](https://mdxjs.com) ([source](https://togithub.com/mdx-js/mdx)) | [`1.6.21` -> `1.6.22`](https://renovatebot.com/diffs/npm/@mdx-js%2freact/1.6.21/1.6.22) | [![age](https://badges.renovateapi.com/packages/npm/@mdx-js%2freact/1.6.22/age-slim)](https://docs.renovatebot.com/merge-confidence/) | [![adoption](https://badges.renovateapi.com/packages/npm/@mdx-js%2freact/1.6.22/adoption-slim)](https://docs.renovatebot.com/merge-confidence/) | [![passing](https://badges.renovateapi.com/packages/npm/@mdx-js%2freact/1.6.22/compatibility-slim/1.6.21)](https://docs.renovatebot.com/merge-confidence/) | [![confidence](https://badges.renovateapi.com/packages/npm/@mdx-js%2freact/1.6.22/confidence-slim/1.6.21)](https://docs.renovatebot.com/merge-confidence/) |
```

## Live example of fix:

| Package | Change | Package age | Adoption rate | Passing tests | Confidence level |
| :---: | :---: | :---: | :---: | :---: | :---: |
| [@mdx-js/react](https://mdxjs.com) ([source](https://togithub.com/mdx-js/mdx)) | [`1.6.21` -> `1.6.22`](https://renovatebot.com/diffs/npm/@mdx-js%2freact/1.6.21/1.6.22) | [![age](https://badges.renovateapi.com/packages/npm/@mdx-js%2freact/1.6.22/age-slim)](https://docs.renovatebot.com/merge-confidence/) | [![adoption](https://badges.renovateapi.com/packages/npm/@mdx-js%2freact/1.6.22/adoption-slim)](https://docs.renovatebot.com/merge-confidence/) | [![passing](https://badges.renovateapi.com/packages/npm/@mdx-js%2freact/1.6.22/compatibility-slim/1.6.21)](https://docs.renovatebot.com/merge-confidence/) | [![confidence](https://badges.renovateapi.com/packages/npm/@mdx-js%2freact/1.6.22/confidence-slim/1.6.21)](https://docs.renovatebot.com/merge-confidence/) |

## Picture of badges not center aligned:

![badge-misaligned](https://user-images.githubusercontent.com/34918129/100885690-3b15b200-34b3-11eb-8d0e-1d821805bc1f.png)

**Comment by rarkins** [comment]
Great finding!

We serve the badges on all platforms so need to determine:
- do the others support this, or it looks wrong?
- should we just massage tables in general to centered in the github platform code? (where we perform other platform-specific markdwon massaging)

And then the other part of what you've found is whether we should use deliberately more verbose column headings in order to get more predictability for badge sizes.

**Comment by HonkingGoose** [comment]
# Platform support for Markdown center aligned tables

| Platform | Supports Markdown center tables |
| --- | :---: |
| GitHub | Yes |
| GitLab | Yes |
| Azure DevOps | Yes |
| Bitbucket | No |

## GitHub supports centering Markdown tables

Quote from [GitHub Flavored Markdown Spec, 4.10 Tables](https://github.github.com/gfm/#tables-extension-):

> 4.10 Tables (extension) 
>
> GFM enables the `table` extension, where an additional leaf block type is available.
>
> The delimiter row consists of cells whose only content are hyphens (`-`), and optionally, a leading or trailing colon (`:`), or both, to indicate left, right, or center alignment respectively.

## GitLab supports centering Markdown tables

Quote from [GitLab Markdown guide, tables](https://docs.gitlab.com/ee/user/markdown.html#tables):

> Additionally, you can choose the alignment of text within columns by adding colons (`:`) to the sides of the “dash” lines in the second row.

## Azure DevOps supports centering Markdown tables:

[Azure DevOps, Markdown Guidance, tables](https://docs.microsoft.com/en-us/azure/devops/project/wiki/markdown-guidance?view=azure-devops#tables)

> Use colons (`:`) when dividing the header and body of tables to specify column alignment (left, center, right)

## BitBucket uses CommonMark tables only, so cannot use centered tables

[Bitbucket server, markdown syntax guide](https://confluence.atlassian.com/bitbucketserver/markdown-syntax-guide-776639995.html#Markdownsyntaxguide-Tables)

**Comment by rarkins** [comment]
Here's a screenshot example of a table with multiple rows:

![image](https://user-images.githubusercontent.com/6311784/100890494-9b5b2280-34b8-11eb-9a2e-4f79750e3cdc.png)

Here's how it looks all center-aligned:

![image](https://user-images.githubusercontent.com/6311784/100890794-ea08bc80-34b8-11eb-9651-b4e96f992c73.png)

I wasn't sure about the Package column, but I think it looks ok?

**Comment by HonkingGoose** [comment]
I think the _Package_ column should be left-aligned.

Also the badge content itself can grow, if _Age_ becomes say `1 month` that whole column gets out of whack with the others that might have `20 days` still. Same problem with Confidence, as that can have different sizes as well (low, medium, high, etc.).

Maybe you're better of doing the table in plain HTML and then add some CSS Grid magic to align everything nicely.

**Comment by rarkins** [comment]
Most platforms don't allow any custom CSS, so that's not an option.

**Comment by HonkingGoose** [comment]
> Most platforms don't allow any custom CSS, so that's not an option.

But you can embed HTML in a Markdown file, and then add inline CSS maybe?

```markdown
This is a regular paragraph.

<table>
    <tr>
        <td style="center-align: true, or whatever works in CSS">Foo</td>
    </tr>
</table>

This is another regular paragraph.
```

**Comment by HonkingGoose** [comment]
> And then the other part of what you've found is whether we should use deliberately more verbose column headings in order to get more predictability for badge sizes.

This will work for most cases, but **_word-wrap_** is causing problems. In some cases the words are being put below each other in the column, that causes the column to shrink again...

**Comment by rarkins** [comment]
> But you can embed HTML in a Markdown file, and then add inline CSS maybe?

I believe GitHub and probably others strip out inline CSS.

So maybe we skip center-aligning so that the badges left-align with each other?

That leaves column width remaining.

The Age badge is probably narrow because the column heading is only 3 characters and the badge itself is not wide either. Another possibility could be to add invisible `&nbsp;` characters like `&nbsp;&nbsp;Age&nbsp;&nbsp;` but we need to check how that appears in each platform.

I think Passing looks too small because the badge itself is wide due to the text "low data". An alternative would be to use the ⏳ symbol(s) in place. Although it's less intuitive than "low data", if the badge links to documentation explaining that's what it means then people only really need to learn that once.

**Comment by HonkingGoose** [comment]
> I believe GitHub and probably others strip out inline CSS.

Ok, inline CSS is out then. 😄 

> So maybe we skip center-aligning so that the badges left-align with each other?

If we cannot be consistent for all platforms, it's probably better to just use the default left-align, and accept that some combinations of badges can look a bit weird in a table with multiple rows.

> An alternative would be to use the ⏳ symbol(s) in place.

That's a possibility to fix the Passing badges,, but it's not very descriptive on its own without reading the docs.
So I'd prefer to keep the text in there.

I hope that the `&nbsp;` thing works, as that might be a good way to make the columns a bit bigger, so that the badge size is preserved.
